Add abi.encodeError

Amxx opened the pull request:
Fixes #14287

TODO:
- [ ] code deducpliation between `typeCheckABIEncodeCallFunction` and `typeCheckABIEncodeErrorFunction` ?
  - I'm not sure if there is a clean way of doing that. Considering there is already a lot of code duplication in the codebase, maybe that is ok.
- [x] tests
- [ ] review

Amxx committed:
Add abi.encodeError

Amxx committed:
deduplicate errors

Amxx committed:
testing

nikola-matic commented on test/libsolidity/syntaxTests/abiEncoder/abi_encodeError_free_func.sol:
```diff
@@ -0,0 +1,7 @@
+    abi.encodeError(g, (1));
+}
+// ----
+// TypeError 3510: (56-57): Expected an error type. Cannot use functions for abi.encodeError.
```
Needs another case for library functions.

nikola-matic reviewed (commented):
Needs some semantic tests as well. A changelog entry as well.

r0qs commented:
This CI error: https://app.circleci.com/pipelines/github/ethereum/solidity/33577/workflows/8c3156a6-35f8-4aeb-b6e3-18f80f4cb6cf/jobs/1512784 is because the test file name is wrong, see: `test/libsolidity/syntaxTests/specialFunctions/encodeCall_fail_args_internal_function_pointer_for_uint copy.sol`. There is a `_copy` in the filename of this test https://github.com/ethereum/solidity/pull/14974/files#diff-41100eaa90ef551806d2098a1c4f80fe0f86d9587edbfb6f1161c6d543f1f92dR1.

Amxx commented:
> Needs some semantic tests as well. A changelog entry as well.

Yes, i'm still working on the testing part.

Amxx committed:
fix test file name

Amxx committed:
add semanticTests - viaIr not working

Amxx committed:
fix IR generator

Amxx commented:
Apparently, in the semantic tests, the returned `bytes memory` are processed in chunks of 32 bytes. So when the buffer has `length ≡ 4 mod 0x20`, we need to pad the expected output with zeros (28 bytes worth of zero exactly)

Is that supposed to happen?

cameel commented:
Could be another instance of this: https://github.com/ethereum/solidity/issues/13989
I think that soltest is just not padding byte arrays correctly.

Amxx committed:
simplify

Amxx committed:
add changelog entry

matheusaaguiar commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -2445,6 +2451,131 @@ void TypeChecker::typeCheckABIEncodeCallFunction(FunctionCall const& _functionCa
+	if (tupleType)
+	{
+		if (TupleExpression const* argumentTuple = dynamic_cast<TupleExpression const*>(arguments[1].get()))
+			callArguments = decltype(callArguments){argumentTuple->components().begin(), argumentTuple->components().end()};
```
```suggestion
			callArguments = {argumentTuple->components().begin(), argumentTuple->components().end()};
```
I think this is better. `std::initializer_list` gets converted automatically.

matheusaaguiar commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -2227,12 +2228,17 @@ void TypeChecker::typeCheckABIEncodeFunctions(
 		typeCheckABIEncodeCallFunction(_functionCall);
 		return;
 	}
+	else if (_functionType->kind() == FunctionType::Kind::ABIEncodeError)
```
```suggestion
	if (_functionType->kind() == FunctionType::Kind::ABIEncodeError)
```

matheusaaguiar commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -2445,6 +2451,131 @@ void TypeChecker::typeCheckABIEncodeCallFunction(FunctionCall const& _functionCa
+				msg += " Cannot use special function.";
+		}
+
+		m_errorReporter.typeError(3510_error, arguments[0]->location(), {}, msg);
```
```suggestion
		m_errorReporter.typeError(3510_error, arguments[0]->location(), msg);
```
We also have one overload which accepts just one location.

matheusaaguiar reviewed (commented):
This needs documentation entries in https://docs.soliditylang.org/en/v0.8.23/cheatsheet.html and https://docs.soliditylang.org/en/v0.8.23/units-and-global-variables.html#abi-encoding-and-decoding-functions (at least, there may be other places, but those are the main ones I guess).
I think the implementation is good, just left some minor suggestions. However, I think we could avoid the duplicated code with some `if`s and ternary conditions and have it all in one `typeCheckABIEncodeCallFunctionOrError` (or something like that), but I don't consider that should impede this PR from being merged.

Amxx committed:
Merge branch 'develop' into feature/abi.encodeError

Amxx committed:
Apply suggestions from code review

Co-authored-by: matheusaaguiar <95899911+matheusaaguiar@users.noreply.github.com>

Amxx committed:
documentation

Amxx committed:
fix tests after merge

Amxx committed:
merge typeCheckABIEncodeCallFunction and typeCheckABIEncodeErrorFunction
into a single typeCheckABIEncodeCallFunctionOrError

Amxx committed:
re-apply suggestion for PR review

Amxx commented:
@matheusaaguiar 
- I added some references in the files you mentionned
- I fixed the merge conflicts and fixed what needed to be fixed
- I also merged previous functions into `typeCheckABIEncodeCallFunctionOrError` (added the type object and some switches)

matheusaaguiar commented on docs/cheatsheet.rst:
```diff
@@ -26,6 +26,8 @@ ABI Encoding and Decoding Functions
 - ``abi.encodeWithSignature(string memory signature, ...) returns (bytes memory)``: Equivalent
   to ``abi.encodeWithSelector(bytes4(keccak256(bytes(signature))), ...)``
+- ``abi.encodeError(error errorPointer, (...)) returns (bytes memory)``: ABI-encodes the revert data for ``errorPointer`` with the arguments found in the
+  tuple. Performs a full type-check, ensuring the types match the error definition. Results equals ``abi.encodeWithSelector(errorPointer.selector, (...))``
```
```suggestion
  tuple. Performs a full type-check, ensuring the types match the error definition. Results equals ``abi.encodeWithSelector(errorPointer.selector, ...)``
```
This a minor imprecision that was fixed recently in the previous entry for `abi.encodeCall` :)
~~I will already commit this suggestion via github interface;~~
Actually, there were another instances which I already fixed via a commit.

matheusaaguiar committed:
fixup! correct minor imprecision

matheusaaguiar commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -2316,17 +2335,32 @@ void TypeChecker::typeCheckABIEncodeCallFunction(FunctionCall const& _functionCa
+				m_errorReporter.typeError(
+					5512_error,
+					arguments.front()->location(),
+					"Expected first argument to be a custom error, not \"" +
```
I was trying to think of some test cases and when I try something like:
```solidity
error E(uint a, uint b);
contract C {
    function g() public view returns (bytes memory) {
        bytes memory b1 = abi.encodeError(E(1, 2), (1, 2));
        return b1;
    }
}
```
The following error message is generated:
`Expected first argument to be a custom error, not "error".`
Which is a bit weird. Maybe we should be more precise and change to `[...] custom error name [...]` ?
I think we should have a test like this anyway.

matheusaaguiar reviewed (approved):
I just have one nitpick left, but overall looks good to me.
Also, I forgot to mention that we use `rebase`  rather than `merge` and this needs the former.

ekpyron commented:
Pending the discussion in https://github.com/ethereum/solidity/issues/14287#issuecomment-2096501517, we decided to close this PR for now. But feel free to jump in there (or join one of our calls) and make the case for this approach and we can potentially revisit and revive this PR!
But as a general policy, we should discuss and confirm the approach before implementations, to be sure we're on the same page about it.

Amxx commented:
@ekpyron [This change to ERC-7751](https://ethereum-magicians.org/t/erc-7751-wrapping-of-bubbled-up-reverts/20740/3?u=amxx), if it was to happen (note, I'm a co-author of the ERC, so not unlikelly), would create a significant usecase of `abi.encodeError`.

Depending on the feedback this gets, I'd like to reconsider this feature.
